src/Synapse/Synapse/Common/AzureSessionCredential.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Microsoft.Azure.Commands.Common.Authentication;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using Microsoft.Azure.Commands.ResourceManager.Common.Properties;
+
+namespace Microsoft.Azure.Commands.Synapse.Common
+{
+    public delegate void DebugLogWriter(string log);
+    public class AzureSessionCredential : TokenCredential
+    {
+        public AzureSessionCredential(IAzureContext DefaultContext, DebugLogWriter logWriter = null)
+        {
+            if (DefaultContext == null || DefaultContext.Account == null)
+            {
+                throw new InvalidOperationException(Resources.ContextCannotBeNull);
+            }
+            if (logWriter != null)
+            {
+                this.debugLogWriter = logWriter;
+            }
+
+            IAccessToken accessToken1 = AzureSession.Instance.AuthenticationFactory.Authenticate(
+               DefaultContext.Account,
+               EnsureSynapseOAuthAudienceSet(DefaultContext.Environment),
+               DefaultContext.Tenant.Id,
+               null,
+               ShowDialog.Never,
+               null,
+               SynapseOAuthEndpointResourceKey);
+            accessToken =  accessToken1;
+        }
+
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            AccessToken token;
+            accessToken.AuthorizeRequest((tokenType, tokenValue) =>
+            {
+                token = new AccessToken(tokenValue, DateTimeOffset.UtcNow);
+            });
+            if (this.debugLogWriter != null)
+            {
+                this.debugLogWriter("[" + DateTime.Now.ToString() + "] GetToken: " + token.Token);
+            }
+            return token;
+        }
+
+        public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            AccessToken token;
+            accessToken.AuthorizeRequest((tokenType, tokenValue) =>
+            {
+                token = new AccessToken(tokenValue, DateTimeOffset.UtcNow);
+            });
+
+            if (this.debugLogWriter != null)
+            {
+                this.debugLogWriter("[" + DateTime.Now.ToString() + "] GetTokenAsync: " + token.Token);
+            }
+            return new ValueTask<AccessToken>(token);
+        }
+
+        private IAzureEnvironment EnsureSynapseOAuthAudienceSet(IAzureEnvironment environment)
+        {
+            if (environment != null)
+            {
+                if (!environment.IsPropertySet(SynapseOAuthEndpointResourceKey))
+                {
+                    environment.SetProperty(SynapseOAuthEndpointResourceKey, SynapseOAuthEndpointResourceValue);
+                }
+            }
+
+            return environment;
+        }
```

Comment:
The environment related change should be in another repo https://github.com/Azure/azure-powershell-common, you may search related code for https://github.com/Azure/azure-powershell-common/blob/master/src/Authentication.Abstractions/AzureEnvironmentConstants.cs#L118..L121 as reference.
BTW, does synapse have same endpoint for different azure cloud environments? E.g. UsGov, GermanGov environments.

Reply:
@erich-wang , we already added environment related config in azure-powershell-common: https://github.com/Azure/azure-powershell-common/blob/master/src/Authentication.Abstractions/AzureEnvironmentConstants.cs#L194-L195 . We added Synapse related endpoints as extended properties. But this comment makes sense to me. @wonner , could you please try removing this method and authenticate as below (line 27)?

``` csharp
            IAccessToken accessToken1 = AzureSession.Instance.AuthenticationFactory.Authenticate(
               DefaultContext.Account,
               DefaultContext.Environment,
               DefaultContext.Tenant.Id,
               null,
               ShowDialog.Never,
               null,
               AzureEnvironment.ExtendedEndpoint.AzureSynapseAnalyticsEndpointResourceId);
```

Reply:
@erich-wang for your 2nd question. Currently Synapse only have endpoints for Global cloud and China cloud. Other national clouds are not supported.